oura-library/src/main/kotlin/org/radarbase/oura/request/OuraRequestGenerator.kt

```diff
@@ -0,0 +1,143 @@
+package org.radarbase.oura.request
+
+import com.fasterxml.jackson.core.JsonFactory
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
+import okhttp3.Response
+import okhttp3.ResponseBody
+import org.radarbase.oura.converter.TopicData
+import org.radarbase.oura.route.OuraRouteFactory
+import org.radarbase.oura.route.Route
+import org.radarbase.oura.user.User
+import org.radarbase.oura.user.UserRepository
+import org.slf4j.LoggerFactory
+import java.time.Duration
+import java.time.Instant
+import kotlin.streams.asSequence
+
+class OuraRequestGenerator @JvmOverloads
+constructor(
+    private val userRepository: UserRepository,
+    private val defaultQueryRange: Duration = Duration.ofDays(15),
+    private val ouraOffsetManager: OuraOffsetManager,
+    public val routes: List<Route> = OuraRouteFactory.getRoutes(userRepository),
+) : RequestGenerator {
+
+    private val userNextRequest: MutableMap<String, Instant> = mutableMapOf()
+
+    public var nextRequestTime: Instant = Instant.MIN
+
+    private val shouldBackoff: Boolean
+        get() = Instant.now() < nextRequestTime
+
+    override fun requests(user: User, max: Int): Sequence<RestRequest> {
+        return if (user.ready()) {
+            routes.asSequence()
+                .flatMap { route ->
+                    return@flatMap generateRequests(route, user)
+                }
+                .takeWhile { !shouldBackoff }
+        } else {
+            emptySequence()
+        }
+    }
+
+    override fun requests(route: Route, max: Int): Sequence<RestRequest> {
+        return userRepository
+            .stream()
+            .flatMap { user ->
+                return@flatMap generateRequests(route, user)
+            }
+            .takeWhile { !shouldBackoff }
+    }
+
+    override fun requests(route: Route, user: User, max: Int): Sequence<RestRequest> {
+        return if (user.ready()) {
+            return generateRequests(route, user)
+        } else {
+            emptySequence()
+        }
+    }
+
+    fun generateRequests(route: Route, user: User): Sequence<RestRequest> {
+        val offset = ouraOffsetManager.getOffset(route, user)
+        val startDate = user.startDate
+        val startOffset: Instant = if (offset == null) {
+            logger.debug("No offsets found for $user, using the start date.")
+            startDate
+        } else {
+            logger.debug("Offsets found in persistence.")
+            offset.offset.coerceAtLeast(startDate)
+        }
+        val endDate = user.endDate
+        if (endDate <= startOffset) return emptySequence()
+        val endTime = (startOffset + defaultQueryRange).coerceAtMost(endDate)
+        return route.generateRequests(user, startOffset, endTime)
+    }
+
+    fun handleResponse(req: RestRequest, response: Response): List<TopicData> {
```

Comment:
I was wondering if we should return a custom sealed Result class here. Something like

```kt
sealed interface OuraError

sealed class OuraErrorBase(val message: String, val cause: Exception? = null): Error {
...
}

sealed class OuraRateLimitError: OuraErrorBase(...) {
...
}

sealed class OuraResult<out T : Any> {
    data class Success<out T : Any>(val value: T) : OuraResult<T>()
    data class Error(val error: OuraError) : OuraResult<Nothing>()
    object Loading : OuraResult <Nothing>() // maybe not needed in this case

}

```

Then downstream clients can check

```kt

when (result) {
is OuraResult.Success -> // Do something
is OuraResult.Error -> {
    when (result.error) {
          is OuraRateLimitError -> // Do something
           ......
    }
....
}

}
```

You also don't need the `else` clause in the above since the subtypes of a sealed class are fixed at compile time.